### PR TITLE
Fix R2 wildcard etag parsing

### PR DESCRIPTION
**Pull request by harshal317**
This commit addresses issue #2572

### CODE REVIEW

File: src/workerd/api/r2-bucket.c++ [hunk 1]
```diff
@@ -935,10 +935,19 @@ kj::Array<R2Bucket::Etag> parseConditionalEtagHeader(kj::StringPtr condHeader,
   }
 }
 
-kj::Array<R2Bucket::Etag> buildSingleStrongEtagArray(kj::StringPtr etagValue) {
-  struct R2Bucket::StrongEtag etag = {.value = kj::str(etagValue)};
+kj::Array<R2Bucket::Etag> buildSingleEtagArray(kj::StringPtr etagValue) {
+  auto value = kj::str(etagValue);
```
Comment (jasnell) [inline src/workerd/api/r2-bucket.c++:939]: Creating the copy here shouldn't be necessary.

File: src/workerd/api/r2-test.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    // Assert it's the correct version
+    assert(jsonRequest.version = 1)
+
+    if (request.method === "PUT") {
+      assert(jsonRequest.method === "put")
+
+      if (jsonRequest.object === "onlyIfStrongEtag") {
+        assert(jsonRequest.onlyIf.etagMatches.length === 1)
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:42]: Since you're using `node:assert`, it is better to use `assert.strictEqual(...)` here instead of `assert(...)`

Comment (jasnell) [reply src/workerd/api/r2-test.js:42]: You could also probably greatly simplify things using `assert.deepStrictEqual(...)` but that's only a guess.

File: src/workerd/api/r2-test.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    const metadataSize = parseInt(metadataSizeString)
+    assert(!Number.isNaN(metadataSize))
+
+    const reader = request.body.getReader({ mode: 'byob' })
+    const jsonArray = new Uint8Array(metadataSize)
+    const { value } = await reader.readAtLeast(metadataSize, jsonArray)
+    const jsonRequest = JSON.parse(new TextDecoder().decode(value))
+
+    // Currently not using the body in these test so I'm going to just discard
+    while (true) {
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:26]: You can simplify this loop to just...

```js
for await (const _ of request.body) {}
```

And eliminate creating the reader

File: src/workerd/api/r2-test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+export default {
+  // Handler for HTTP request binding makes to R2
+  async fetch(request, env, ctx) {
+    // We only expect PUT/Get
+    assert(request.method === "PUT" || request.method === "GET")
+
+    // Each request should have a metadata size header indicating how much
+    // we should read to understand what type of request this is
+    const metadataSizeString = request.headers.get("cf-r2-metadata-size")
+    assert(metadataSizeString !== null)
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:16]:
```js
asssert.notStrictEqual(metadataSizeString, null);
```

File: src/workerd/api/r2-test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+import assert from 'node:assert';
+
+export default {
+  // Handler for HTTP request binding makes to R2
+  async fetch(request, env, ctx) {
+    // We only expect PUT/Get
+    assert(request.method === "PUT" || request.method === "GET")
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:11]:
```js
assert(['GET', 'PUT'].includes(request.method));
```

File: src/workerd/api/r2-test.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+      if (jsonRequest.object === "onlyIfWildcard") {
+        assert(jsonRequest.onlyIf.etagMatches.length === 1)
+        assert(jsonRequest.onlyIf.etagDoesNotMatch.length === 1)
+
+        assert(jsonRequest.onlyIf.etagMatches[0].type === "wildcard")
+        assert(jsonRequest.onlyIf.etagDoesNotMatch[0].type === "wildcard")
+      }
+
+      return new Response(new TextEncoder().encode(JSON.stringify({
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:60]: You can simplify this using...

```js
return Response.json({
  name: 'objectKey',
  // ...
});
```

File: src/workerd/api/r2-test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        etagMatches: "strongEtag",
+        etagDoesNotMatch: "strongEtag"
+      }
+    })
+    await env.BUCKET.put("onlyIfWildcard", "content", {
+      onlyIf: {
+        etagMatches: "*",
+        etagDoesNotMatch: "*"
+      }
+    })
```
Comment (jasnell) [inline src/workerd/api/r2-test.js:87]: What is the expectation when someone uses a weak etag? e.g. `W/"foo"`

A test verifying the expectation would help round this out.

Comment (harshal317) [reply src/workerd/api/r2-test.js]: I don't think we need to support weak eTags here, and we hadn't been. We do support them through the HTTP headers to follow the spec but a weak eTag on etagMatches is never going to be true and we always use weak comparison on etagDoesNotMatch so regardless of strong/weak we treat it the same. 

Maybe worth updating the docs to clarify this.

File: src/workerd/api/r2-test.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    // Assert it's the correct version
+    assert((jsonRequest.version = 1));
+
+    if (request.method === "PUT") {
+      assert(jsonRequest.method === "put");
+
+      if (jsonRequest.object === "onlyIfStrongEtag") {
+        assert(jsonRequest.onlyIf.etagMatches.length === 1);
```
Comment (anonrig) [inline src/workerd/api/r2-test.js:42]: If you use `assert.strictEqual`, we will have better debugging if someone breaks this test.

File: src/workerd/api/r2-test.js (hunk 2 above)
Comment (harshal317) [reply src/workerd/api/r2-test.js]: That's a lot cleaner, thanks for the tip!

File: src/workerd/api/r2-test.js (hunk 8 above)
Comment (harshal317) [reply src/workerd/api/r2-test.js]: Changed to deepStrictEqual

**Review by jasnell — APPROVED** [review]

**Review by anonrig — APPROVED** [review]

**Comment by harshal317** [comment]
Oops thought maybe clicking close would give me an option to merge.

**Comment by harshal317** [comment]
@jasnell @anonrig should I get a review from anyone else? If not, could one of you help me with merging this

**Comment by jasnell** [comment]
Have you run an internal CI job with this update to verify everything works internally? If so, this should be ready to land. If not, let's do that first.

**Pull request merged**